qt: Add privacy to the Overview page

hebasto opened the pull request:
This PR allows to hide/reveal values on the Overviewpage by checking/unchecking Menu->Settings-> Mask Values

Closes #16407 

Privacy mode is OFF (the default behavior):
![Screenshot from 2020-01-02 15-08-28](https://user-images.githubusercontent.com/32963518/71669074-28ab6980-2d74-11ea-8e54-4973aa307192.png)

Privacy mode is ON:
![Screenshot from 2020-01-02 15-10-23 cropped](https://user-images.githubusercontent.com/32963518/71669082-2d701d80-2d74-11ea-9df5-d4acc4982dbe.png)

practicalswift commented:
@hebasto Would you mind posting before/after screenshots?

hebasto commented:
@practicalswift 
> Would you mind posting before/after screenshots?

Done.

MarcoFalke commented on src/qt/bitcoinamountlabel.cpp:
```diff
@@ -0,0 +1,36 @@
+{
+    if (privacy && cache.isEmpty()) {
+        cache = text();
+        QLabel::setText("h-i-d-d-e-n");
```
I think a fence of `#` translates better.

Couldn't this use `BitcoinUnits::format(unit, 0).replace("0", "#") + unit.shortName()` (or whatever is used right now), so that the UI doesn't reflow the columns on every toggle?

MarcoFalke replied:
![Screenshot from 2019-07-22 08-04-35](https://user-images.githubusercontent.com/6399679/61631080-66e61e00-ac57-11e9-9143-34c0965c3f89.png)

hebasto replied:
> I think a fence of # translates better.

Done.

> ... so that the UI doesn't reflow the columns on every toggle?

This requires a monospaced font.

hebasto replied:
> This requires a monospaced font.

Done.

laanwj commented:
Concept ACK, definitely prefer the `#` to `h-i-d-d-e-n`

hebasto commented:
@MarcoFalke 
> I think a fence of `#` translates better.

@laanwj 
> definitely prefer the `#` to `h-i-d-d-e-n`

The screenshot on OP has been updated.

achow101 commented:
Can it use a fixed number/string of `#`s instead of replacing each character with `#`? That way we don't leak any information about the number of coins. For example, right now, in your screenshot, we can see that you have at least 1000 coins and at least 100 available to spend.

jonasschnelli commented:
Thanks for working on this!
Tested a bit. Works.

Not sure if clicking on the actual amount label is the right way to activate this (may confuse user who lacks understanding the "hidden"-concept).

Plus, should it not also cover the "Recent transactions"-list amounts?

I would say ideally, once enabled, the values in the transaction list should also be "hidden".

promag commented:
Concept ACK. Not so sure about the approach as it requires too much wiring. I'm testing a different approach and I'll give feedback if it's worth it.

promag commented:
> Not sure if clicking on the actual amount label is the right way to activate this 

Agree with @jonasschnelli, I initially thought of having a toggle somewhere (in the status bar for instance) as well a menu option. Something for a follow up.

hebasto commented:
@achow101 
> Can it use a fixed number/string of `#`s instead of replacing each character with `#`? That way we don't leak any information about the number of coins. For example, right now, in your screenshot, we can see that you have at least 1000 coins and at least 100 available to spend.

Done. The screenshot on OP has been updated.

hebasto commented:
@jonasschnelli Thank you for your review.

> Not sure if clicking on the actual amount label is the right way to activate this (may confuse user who lacks understanding the "hidden"-concept).

It seems very intuitive ;) (also there is an example of usage in Wasabi)
To further improvements (on the following PR), a toggle could be added (as @promag [suggested](https://github.com/bitcoin/bitcoin/pull/16432#issuecomment-513814481)).

> Plus, should it not also cover the "Recent transactions"-list amounts?
> I would say ideally, once enabled, the values in the transaction list should also be "hidden".

Agree. I'd leave this improvement for the following PRs.

kristapsk commented:
> Not sure if clicking on the actual amount label is the right way to activate this (may confuse user who lacks understanding the "hidden"-concept).

That's the way "lurking wife mode" of Wasabi Wallet, which is basically the same, is activated. I think Bitcoin is already hard enough for newcomers, so unnecessary reinventing the wheel should be avoided, if possible, so that users get similar experience using different wallets. Unless there are clear benefits in doing so, of course. But probably some message about privacy / stealth mode being activated and how to turn it back off would not hurt.

kristapsk commented on src/qt/bitcoinamountlabel.cpp:
```diff
@@ -0,0 +1,47 @@
+        integer_part = QString(integer_part.size() - 1, ' ') + mask;
+        QString fractional_part = parts.last();
+        fractional_part = fractional_part.replace(QRegularExpression("\\d"), mask);
+        QLabel::setText(integer_part + '.' + fractional_part);
```
Just looked at how Wasabi did this and they just replace any amounts, addresses, labels everywhere with a fixed number of #'s. IMO "#########" even looks better than "#.########" and it's less lines of code, all this block could be just replaced with
```
cache = text();
QLabel::setText("#########");`
```

hebasto replied:
Unfortunately, this will cause the UI to reflow the columns on every toggle (see @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/16432#discussion_r305807357)).

fanquake commented:
Concept ACK

hebasto commented:
Fixed two bugs:
- sat units (no fractional part) handling
- unit switching handling

fanquake commented:
Had a quick test. Functionality works as described. I might just be seeing things, but the monospace font looks a bit weird no macOS, don't think that has to be any sort of blocker though.

@promag Are you still working on a different approach, or happy with this PR now?

master:
![master](https://user-images.githubusercontent.com/863730/61767561-406ee100-ae17-11e9-9b0d-ec8295057ca1.png)

This PR:
![16432-values](https://user-images.githubusercontent.com/863730/61767565-45339500-ae17-11e9-8c35-91d14d4a3533.png)
![16432-hidden](https://user-images.githubusercontent.com/863730/61767567-4795ef00-ae17-11e9-9709-9a33731e86e2.png)

darosior commented:
ACK 22c16fd3e19437db187a514ba043af635135b719
I readed the code and I tested it. I think it's a great addition.

promag commented:
I've tried alternatives - basically tried filtering paint events if the widget had amounts or even extend QStyle - but didn't like at the end.

Currently I think this PR can be improved, I'll review shortly.

jonasschnelli commented:
Looks like the build is failing:
```
qt/bitcoinamountlabel.cpp:26:47: error: 'QRegularExpression' is an incomplete type
    fractional_part = fractional_part.replace(QRegularExpression("\\d"), mask);
                                              ^
/home/ubuntu/src/depends/x86_64-apple-darwin14/include/QtCore/qstring.h:80:7: note: forward declaration of 'QRegularExpression'
class QRegularExpression;
```

https://bitcoinbuilds.org/index.php?build=669

hebasto commented:
@jonasschnelli 
> Looks like the build is failing: ...

Rebased and fixed by partially reverting the 248e22bbc0d7bc40ae3584d53a18507c46b0e553 commit (#16386).

jonasschnelli commented:
Works as intended. I like the concept but think it should be implemented more radical (maybe in follow up PRs?).

* The privacy mode won't hide recent transaction values
* For users not aware of the privacy mode, it's unclear why the numbers are now # and how to bring it back,... a quick disappearing info-element (statusbar?) would make things much more understandable for users new to this concept (or a toggle element in the icon section of the status bar?).
* I think all amount values should be masqueraded, especially the ones in the transaction list.

luke-jr commented on depends/packages/qt.mk:
```diff
@@ -87,7 +87,6 @@ $(package)_config_opts += -no-feature-printdialog
 $(package)_config_opts += -no-feature-printer
 $(package)_config_opts += -no-feature-printpreviewdialog
 $(package)_config_opts += -no-feature-printpreviewwidget
-$(package)_config_opts += -no-feature-regularexpression
```
It seems a bit excessive to add regex support just for this...

laanwj replied:
Would agree here. If you really need complex regexps then I don't mind adding the dep, but it seems overkill for this.

laanwj replied:
OTOH #17213 seems to need qt-regexp as well.

hebasto replied:
Reverted back.

luke-jr reviewed (changes requested):
Rather than cloaking the text we want to hide (which seems problematic so long as we base it on the real hidden value), how about simply setting the background colour to the text colour?

konez2k commented:
This should work without regexp

```
static QString cloak(QString& text)
{
    QStringList parts = text.split(' ');
    QStringList amounts = parts[0].split('.');
    for (int i = 0; i < amounts.size(); i++)
        amounts[i].fill('#');
    return amounts.join('.') + ' ' + parts[1];
}
```

JeremyCrookshank commented:
Are you still working on this @hebasto. This PR is definitely of interest for me and would love to help/takeover it if you aren't

hebasto commented:
@kristapsk 
> But probably some message about privacy / stealth mode being activated and how to turn it back off would not hurt.

@jonasschnelli 
> For users not aware of the privacy mode, it's unclear why the numbers are now # and how to bring it back,... a quick disappearing info-element (statusbar?) would make things much more understandable for users new to this concept

A relevant message has been added to the status bar.

---

@luke-jr 
> Rather than cloaking the text we want to hide (which seems problematic so long as we base it on the real hidden value), how about simply setting the background colour to the text colour?

Nice suggestion! Honestly, I don't like the look of some white bars in a black window (in the dark desktop theme) ;)

---

@konez2k 
> This should work without regexp
> 
> ```
> static QString cloak(QString& text)
> {
>     QStringList parts = text.split(' ');
>     QStringList amounts = parts[0].split('.');
>     for (int i = 0; i < amounts.size(); i++)
>         amounts[i].fill('#');
>     return amounts.join('.') + ' ' + parts[1];
> }
> ```

`text.split(' ')` could return more then two parts, e.g., `1 234.56 bits` ;)

Nevertheless, the latest push works without `QRegularExpression`.

---

@JeremyCrookshank 
> This PR is definitely of interest for me and would love to help...

Thank you! Would you mind reviewing this PR?

hebasto commented:
@jonasschnelli 
> Plus, should it not also cover the "Recent transactions"-list amounts?

Done.

hebasto commented:
All comments have been addressed. PR is ready to be re-reviewed ;)

hebasto commented:
Rebased in order to incorporate already merged #15529 and fix MSVC build.

luke-jr commented on src/qt/forms/overviewpage.ui:
```diff
@@ -115,9 +115,10 @@
+             <widget class="BitcoinAmountLabel" name="labelWatchPending">
               <property name="font">
                <font>
+                <family>monospace</family>
```
Why not just make this the default for `BitcoinAmountLabel`?

hebasto replied:
I've tried different approaches, including
```
const QFont font{"monospace"};
QApplication::setFont(font, "BitcoinAmountLabel");
```

Only the current code works as expected. But I'm still open for suggestions.

hebasto replied:
Done.

luke-jr commented:
This kind of interferes with selecting to copy... :/

JeremyCrookshank commented:
- As Luke said it's very hard to copy and paste values with this enabled also


- It's not obvious why they're hidden if a user is still downloading blocks. I think the best solution is to disable this feature by default and only enable through GUI/config setting. 

![image](https://user-images.githubusercontent.com/46864828/68993661-73c7f680-0872-11ea-933a-d42bdb723e79.png)

- It might be better as well to have an icon indicator instead of using the entire status bar when in privacy mode.

![image](https://user-images.githubusercontent.com/46864828/68993994-8d1e7200-0875-11ea-9c06-58bdefe2f395.png)

- I also think it would be a better idea to show the recent transactions but hide the BTC amount of each transaction so you can still see your recent transactions perhaps

- Also do we need to stop this happening on windows if in privacy mode?(the popup shows value)

![image](https://user-images.githubusercontent.com/46864828/68993839-5005b000-0874-11ea-9205-141af31b5076.png)

hebasto commented:
@JeremyCrookshank 

Thank you for your review.

> As Luke said it's very hard to copy and paste values with this enabled also

It seems a bit counter-intuitive to try copy a hidden / masked value.

>  I think the best solution is to disable this feature by default and only enable through GUI/config setting.

I'd prefer to make GUI/config settings related changes in follow-up PRs.

> It might be better as well to have an icon indicator instead of using the entire status bar when in privacy mode.

An icon indicator could be added in follow-up PR. But I'm not agree to replace a status bar message with an icon indicator completely: a text message is (mostly) always clear, an icon needs more cognitive efforts from a user.

> I also think it would be a better idea to show the recent transactions but hide the BTC amount of each transaction so you can still see your recent transactions perhaps

Such implementation details could be improved in follow-up PRs.

> Also do we need to stop this happening on windows if in privacy mode?(the popup shows value)

Good idea, and it could be implemented in follow-up PRs. A full-fledged privacy mode is not the goal of the current PR.

JeremyCrookshank commented:
> @JeremyCrookshank
> 
> Thank you for your review.
> 
> > As Luke said it's very hard to copy and paste values with this enabled also
> 
> It seems a bit counter-intuitive to try copy a hidden / masked value.

It's hard to copy the non masked value. As you have to right click to copy which masks it.

luke-jr commented:
I typically copy values by double-clicking. No right-click needed, but still broken by this feature.

Maybe a checkable menu item is best?

RandyMcMillan commented:
Using an asterisk as the replacement character may be easier on the eyes.

*.******

and conforms with what users may be used to as far as password obfuscation.

hebasto commented:
@JeremyCrookshank 
> It's hard to copy the non masked value. As you have to right click to copy which masks it.

@luke-jr 
> I typically copy values by double-clicking. No right-click needed, but still broken by this feature.

Click-to-hide/reveal buggy feature has been removed.

> Maybe a checkable menu item is best?

Implemented in the latest push e9df42cc7066f089c2c83844893a677920ee5859.

hebasto commented:
@luke-jr 
> Why not just make this the default for `BitcoinAmountLabel`?

Done in the latest push.

MarkLTZ commented:
Nice PR. I think it just miss to hide the transactions page.

hebasto commented:
@MarkLTZ 
> Nice PR. 

Thank you.

> I think it just miss to hide the transactions page.

I believe this improvement could follow this PR.

luke-jr commented on src/qt/bitcoinamountlabel.cpp:
```diff
@@ -0,0 +1,57 @@
+const QChar point{'.'};
+const QChar mask{'#'};
+
+QString cloak(QString value_and_unit)
```
Basing it on the real number is revealing too much info.

hebasto replied:
> Basing it on the real number is revealing too much info.

I'm not sure if I understand you correctly. Mind rewording your concerns?

luke-jr replied:
Your cloaking reveals at least how many whole number digits are in the real amount, by measuring the number of spaces you add.

https://github.com/bitcoin/bitcoin/commit/98ce0b830e505ddede5d56951e4ce7ee1bf5a0c5

hebasto replied:
> Your cloaking reveals at least how many whole number digits are in the real amount, by measuring the number of spaces you add.

Fixed in the latest push.

> [98ce0b8](https://github.com/bitcoin/bitcoin/commit/98ce0b830e505ddede5d56951e4ce7ee1bf5a0c5)

Monospace font style prevents the UI to reflow the columns on every toggle (https://github.com/bitcoin/bitcoin/pull/16432#discussion_r305807357).

luke-jr commented on src/qt/bitcoinamountlabel.h:
```diff
@@ -0,0 +1,34 @@
+    void setPrivacyMode(bool privacy);
+
+private:
+    QString cache{};
```
Should be called `m_cache`

hebasto replied:
Fixed.

luke-jr commented on src/qt/overviewpage.cpp:
```diff
@@ -139,6 +142,15 @@ OverviewPage::OverviewPage(const PlatformStyle *platformStyle, QWidget *parent)
+    connect(this, &OverviewPage::setPrivacyMode, ui->labelWatchAvailable, &BitcoinAmountLabel::setPrivacyMode);
+    connect(this, &OverviewPage::setPrivacyMode, ui->labelWatchPending, &BitcoinAmountLabel::setPrivacyMode);
+    connect(this, &OverviewPage::setPrivacyMode, ui->labelWatchImmature, &BitcoinAmountLabel::setPrivacyMode);
+    connect(this, &OverviewPage::setPrivacyMode, ui->labelWatchTotal, &BitcoinAmountLabel::setPrivacyMode);
```
I wonder if there's a good way to auto-connect these

hebasto replied:
I believe this could follow this PR.

hebasto replied:
This part of code has been removed in the latest push.

hebasto commented:
In the latest push:
- code is much simplified
- monospace font style continues to be used
- a shortcut has been added to simplify testing and/or usage

![Screenshot from 2020-01-02 15-08-28](https://user-images.githubusercontent.com/32963518/71669074-28ab6980-2d74-11ea-8e54-4973aa307192.png)

![Screenshot from 2020-01-02 15-10-23 cropped](https://user-images.githubusercontent.com/32963518/71669082-2d701d80-2d74-11ea-9df5-d4acc4982dbe.png)

Currently, the privacy mode is _not_ inherited when a wallet is added/loaded.

hebasto commented:
The OP has been updated.

@achow101 @jonasschnelli @fanquake @laanwj @luke-jr @MarcoFalke @promag
Would you mind re-reviewing this PR?

achow101 commented:
ACK 3935bce9c28eefe38ea4e0152dbbddc630671d27

Tested change and reviewed.

Not sure if this is possible, but it would be nice if the spaces for the right justify weren't selectable. Is there some other way to achieve the justify using spacers in the ui form?

hebasto commented:
@achow101 
> Not sure if this is possible, but it would be nice if the spaces for the right justify weren't selectable. Is there some other way to achieve the justify using spacers in the ui form?

It is possible, but it requires to overhaul a form layout in a separate PR.

luke-jr commented on src/qt/bitcoinamountlabel.cpp:
```diff
@@ -0,0 +1,25 @@
+
+void BitcoinAmountLabel::setValue(const CAmount& amount, int unit, bool privacy)
+{
+    QString value = BitcoinUnits::format(unit, privacy ? 0 : amount, false, BitcoinUnits::separatorAlways, true);
```
Please use `formatWithUnit`

hebasto replied:
A new signature of `BitcoinUnits::format()` is used here.

luke-jr commented on src/qt/bitcoinamountlabel.cpp:
```diff
@@ -0,0 +1,25 @@
+{
+}
+
+void BitcoinAmountLabel::setValue(const CAmount& amount, int unit, bool privacy)
```
I'm not sure there's a point to a dedicated widget anymore...

hebasto replied:
Agree. Fixed in the latest push.

luke-jr commented on src/qt/bitcoinamountlabel.cpp:
```diff
@@ -0,0 +1,25 @@
+    if (privacy) {
+        value.replace('0', '#');
+    }
+    QLabel::setText(value + QString(" ") + BitcoinUnits::shortName(unit));
```
`QLabel::` is unnecessary

hebasto committed:
qt: Add BitcoinUnits::formatWithPrivacy() function

hebasto committed:
qt: Add privacy feature to Overview page

hebasto commented on src/qt/forms/overviewpage.ui:
```diff
@@ -6,8 +6,8 @@
     <y>0</y>
-    <width>596</width>
-    <height>342</height>
+    <width>798</width>
```
Note to reviewers: these changes in numbers are just Qt Designer artefacts.

hebasto commented:
As @luke-jr rightly [commented](https://github.com/bitcoin/bitcoin/pull/16432#discussion_r362976691):
> I'm not sure there's a point to a dedicated widget anymore...

So the dedicated `BitcoinAmountLabel` widget has been removed.

hebasto commented:
Rebased after #17937 has been merged.

kristapsk commented:
I think good place for icon when this mode is enabled would be near "Balances" and "Recent transactions" titles, at is is already done during sync, which informs that balances may be wrong and recent transactions might not appear. As this privacy mode also impacts the same things. But agree that this could be added as a follow-up PR. Also, more important IMHO would be to work on hiding values in other parts of interface too when this mode is enabled.

kristapsk reviewed (approved):
ACK 08a3048dd703d67f878d16be00a6288a5d27157f (reviewed code and tested).

luke-jr commented:
Hmm, I wonder if just "closing" the wallets in the GUI (but leaving them open for sync/updates) would work best.

hebasto commented:
> Hmm, I wonder if just "closing" the wallets in the GUI (but leaving them open for sync/updates) would work best.

TBH, I do not like how it would look like (from the point of a user).

hebasto commented:
Rebased 08a3048dd703d67f878d16be00a6288a5d27157f -> dba83b9dab919ea9c72f4321612605eb3242d523 due to the conflict with #18278.

hebasto commented:
Rebased dba83b9dab919ea9c72f4321612605eb3242d523 -> 6920e1236b3f7a6b46c78f058aac4a32fcfede9f due to the conflict with #16528.

hebasto commented:
@achow101 @jonasschnelli @luke-jr @promag and @instagibbs (as initiator of #16407)
Mind re-reviewing this PR?

instagibbs commented:
@hebasto huh didn't know this PR existed. Will review.

instagibbs commented:
just from looking at the screenshots, did you consider something like adding a GUI toggle above/near the balances themselves? It's probably something I'd want to toggle on and off pretty fast, yeah?

Concept ACK on the ##### scheme at least

hebasto commented:
> just from looking at the screenshots, did you consider something like adding a GUI toggle above/near the balances themselves? It's probably something I'd want to toggle on and off pretty fast, yeah?

https://github.com/bitcoin/bitcoin/blob/6920e1236b3f7a6b46c78f058aac4a32fcfede9f/src/qt/bitcoingui.cpp#L354

instagibbs commented:
I see, it's a top-level setting. Light testing shows it does what's on the tin. 

tACK https://github.com/bitcoin/bitcoin/pull/16432/commits/6920e1236b3f7a6b46c78f058aac4a32fcfede9f

jonatack commented on src/qt/bitcoingui.cpp:
```diff
@@ -350,6 +350,11 @@ void BitcoinGUI::createActions()
 
+    m_mask_values_action = new QAction(tr("&Mask values"), this);
+    m_mask_values_action->setShortcut(QKeySequence(Qt::CTRL + Qt::SHIFT + Qt::Key_M));
+    m_mask_values_action->setStatusTip(tr("Cloak values on Overview tab"));
```
- s/values on/the values in the|all values in the/ 

- I'd suggest keeping the verb consistent everywhere, e.g. s/Cloak/Mask/

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/16432#issuecomment-633063508).

jonatack commented on src/qt/overviewpage.cpp:
```diff
@@ -152,6 +154,23 @@ void OverviewPage::handleOutOfSyncWarningClicks()
+        ui->listTransactions->setVisible(!m_privacy);
+    }
+
+    const QString status_tip = m_privacy ? tr("PRIVACY mode is activated. To reveal cloaked values uncheck Menu->Settings->Mask values") : "";
```
suggestion:
```diff
- "PRIVACY mode is activated. To reveal cloaked values uncheck Menu->Settings->Mask values"
+ "PRIVACY mode is activated. To unmask the values, uncheck Menu->Settings->Mask values."
```
- adds a missing "the", comma, and period
- replaces "reveal cloaked" with the more consistent, shorter, and simpler "unmask"

(also, maybe consider not screaming :) e.g. s/PRIVACY/Privacy/)

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/16432#issuecomment-633063508).

jonatack commented on src/qt/bitcoinunits.cpp:
```diff
@@ -150,6 +153,15 @@ QString BitcoinUnits::formatHtmlWithUnit(int unit, const CAmount& amount, bool p
+        value.replace('0', '#');
+    }
+    return value + QString(" ") + shortName(unit);
+}
```
micro-nit, can perform one less conditional (with slightly improved readability) by avoiding the embedded ternary:
```diff
-    QString value = format(unit, privacy ? 0 : amount, false, separators, true);
+    QString value;
     if (privacy) {
+        value = format(unit, 0, false, separators, true);
         value.replace('0', '#');
+    } else {
+        value = format(unit, amount, false, separators, true);
     }
```

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/16432#issuecomment-633063508).

jonatack reviewed (commented):
Tested ACK 6920e1236b3f7 with some suggestions if you retouch or for a follow-up. Feel free to ignore the code nits; I'm more concerned with the user-facing ones.

I wonder if it's stealthy enough with the screaming mode activation message at the bottom. For a follow-up, it might be good to add a user option in Settings -> Options -> Display (or Privacy) to allow not displaying it.

All in all, a cool addition :sunglasses:

hebasto commented:
Updated 6920e1236b3f7a6b46c78f058aac4a32fcfede9f -> 206d17e4d7131179e7c5a5e74632806859d7d493 ([pr16432.17](https://github.com/hebasto/bitcoin/commits/pr16432.17) -> [pr16432.18](https://github.com/hebasto/bitcoin/commits/pr16432.18), [diff](https://github.com/hebasto/bitcoin/compare/pr16432.17..pr16432.18)):

- addressed all @jonatack's comments

hebasto commented on src/qt/bitcoinunits.h:
```diff
@@ -72,11 +72,13 @@ class BitcoinUnits: public QAbstractListModel
     static QString formatWithUnit(int unit, const CAmount& amount, bool plussign=false, SeparatorStyle separators=separatorStandard);
     //! Format as HTML string (with unit)
     static QString formatHtmlWithUnit(int unit, const CAmount& amount, bool plussign=false, SeparatorStyle separators=separatorStandard);
+    //! Format as string (with unit) of fixed length to preserve privacy if it set
```
[Updated](https://github.com/bitcoin/bitcoin/pull/16432#issuecomment-633063508).

instagibbs commented:
reACK https://github.com/bitcoin/bitcoin/pull/16432/commits/206d17e4d7131179e7c5a5e74632806859d7d493

jonatack reviewed (commented):
ACK 206d17e reviewed changes (thanks!) since my last review of 6920e12, rebased on current master, built and tested on Debian.

<details><summary>Changes since last review per <code>git diff 6920e12 206d17e</code></summary><p>

```diff
$ git diff 6920e12 206d17e
diff --git a/src/qt/bitcoingui.cpp b/src/qt/bitcoingui.cpp
index 9d695f1215..a2fd231c05 100644
--- a/src/qt/bitcoingui.cpp
+++ b/src/qt/bitcoingui.cpp
@@ -352,7 +352,7 @@ void BitcoinGUI::createActions()
 
     m_mask_values_action = new QAction(tr("&Mask values"), this);
     m_mask_values_action->setShortcut(QKeySequence(Qt::CTRL + Qt::SHIFT + Qt::Key_M));
-    m_mask_values_action->setStatusTip(tr("Cloak values on Overview tab"));
+    m_mask_values_action->setStatusTip(tr("Mask the values in the Overview tab"));
     m_mask_values_action->setCheckable(true);
 
     connect(quitAction, &QAction::triggered, qApp, QApplication::quit);
diff --git a/src/qt/bitcoinunits.cpp b/src/qt/bitcoinunits.cpp
index cf7051cda8..318a6dcbfd 100644
--- a/src/qt/bitcoinunits.cpp
+++ b/src/qt/bitcoinunits.cpp
@@ -156,9 +156,11 @@ QString BitcoinUnits::formatHtmlWithUnit(int unit, const CAmount& amount, bool p
 QString BitcoinUnits::formatWithPrivacy(int unit, const CAmount& amount, SeparatorStyle separators, bool privacy)
 {
     assert(amount >= 0);
-    QString value = format(unit, privacy ? 0 : amount, false, separators, true);
+    QString value;
     if (privacy) {
-        value.replace('0', '#');
+        value = format(unit, 0, false, separators, true).replace('0', '#');
+    } else {
+        value = format(unit, amount, false, separators, true);
     }
     return value + QString(" ") + shortName(unit);
 }
diff --git a/src/qt/bitcoinunits.h b/src/qt/bitcoinunits.h
index 2d0b85126e..5b8c351d73 100644
--- a/src/qt/bitcoinunits.h
+++ b/src/qt/bitcoinunits.h
@@ -77,7 +77,7 @@ public:
     static QString formatWithUnit(int unit, const CAmount& amount, bool plussign=false, SeparatorStyle separators=separatorStandard);
     //! Format as HTML string (with unit)
     static QString formatHtmlWithUnit(int unit, const CAmount& amount, bool plussign=false, SeparatorStyle separators=separatorStandard);
-    //! Format as string (with unit) of fixed length to preserve privacy if it set
+    //! Format as string (with unit) of fixed length to preserve privacy, if it is set.
     static QString formatWithPrivacy(int unit, const CAmount& amount, SeparatorStyle separators, bool privacy);
     //! Parse string to coin amount
     static bool parse(int unit, const QString &value, CAmount *val_out);
diff --git a/src/qt/overviewpage.cpp b/src/qt/overviewpage.cpp
index b80fb635ad..48ad438eb2 100644
--- a/src/qt/overviewpage.cpp
+++ b/src/qt/overviewpage.cpp
@@ -165,7 +165,7 @@ void OverviewPage::setPrivacy(bool privacy)
         ui->listTransactions->setVisible(!m_privacy);
     }
 
-    const QString status_tip = m_privacy ? tr("PRIVACY mode is activated. To reveal cloaked values uncheck Menu->Settings->Mask values") : "";
+    const QString status_tip = m_privacy ? tr("PRIVACY mode is activated. To unmask the values, uncheck Menu->Settings->Mask values.") : "";
     setStatusTip(status_tip);
\```
</p></details>

hebasto commented:
@MarcoFalke Mind re-reviewing this PR?

MarcoFalke commented:
Concept 🅰🅲🅺

promag commented on src/qt/overviewpage.cpp:
```diff
@@ -152,6 +154,23 @@ void OverviewPage::handleOutOfSyncWarningClicks()
+void OverviewPage::setPrivacy(bool privacy)
+{
+    m_privacy = privacy;
+    if (walletModel && walletModel->getOptionsModel()) {
```
This condition looks unnecessary/unrelated, could add a comment maybe?

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/16432#issuecomment-633392179).

promag commented on src/qt/overviewpage.cpp:
```diff
@@ -163,25 +182,25 @@ void OverviewPage::setBalance(const interfaces::WalletBalances& balances)
     m_balances = balances;
     if (walletModel->wallet().isLegacy()) {
         if (walletModel->wallet().privateKeysDisabled()) {
-            ui->labelBalance->setText(BitcoinUnits::formatWithUnit(unit, balances.watch_only_balance, false, BitcoinUnits::separatorAlways));
```
Could we simplify this? Maybe define a `class AmountLabel : public QLabel` and here you would just call `setAmount(...)`. And in `setWalletModel` you could `ui->labelBalance->setOptionsModel`. This way `AmountLabel` would call `setText` when privacy or display unit changes (and amount ofc).

hebasto replied:
Removing a tailored widget was [considered](https://github.com/bitcoin/bitcoin/pull/16432#issuecomment-570813848) as simplification already :)
> As @luke-jr rightly [commented](https://github.com/bitcoin/bitcoin/pull/16432#discussion_r362976691):
> > I'm not sure there's a point to a dedicated widget anymore...
>
> So the dedicated `BitcoinAmountLabel` widget has been removed.

Anyway, let's focus this PR on user-faced changes, and leave internal optimization for followup pulls, if you don't mind :)

promag commented on src/qt/overviewpage.h:
```diff
@@ -49,6 +50,7 @@ public Q_SLOTS:
     ClientModel *clientModel;
     WalletModel *walletModel;
     interfaces::WalletBalances m_balances;
+    bool m_privacy{false};
```
I think this should be in `OptionsModel`. No need to go through bitcoingui -> walletframe -> walletview -> overview.

hebasto replied:
> I think this should be in `OptionsModel`.

In this PR the privacy mode state is not intended to be preserved between consequent `bitcoin-qt` runs, that an `OptionsModel` instance does.

I'd leave the detailed discussion for a followup pull.

promag replied:
It doesn't have to be persisted to be in `OptionsModel`, it's a runtime option only for now.

hebasto replied:
Sorry, going to leave it as is.

promag commented on src/qt/walletframe.cpp:
```diff
@@ -73,6 +74,10 @@ bool WalletFrame::addWallet(WalletModel *walletModel)
     connect(walletView, &WalletView::incomingTransaction, gui, &BitcoinGUI::incomingTransaction);
     connect(walletView, &WalletView::hdEnabledStatusChanged, gui, &BitcoinGUI::updateWalletStatus);
+    connect(gui, &BitcoinGUI::setPrivacy, walletView, &WalletView::setPrivacy);
+    QTimer::singleShot(0, walletView, [this, walletView] {
```
Why singleShot? Anyway, if you follow the OptionsModel then this is not needed.

hebasto replied:
> Why singleShot?

To apply the privacy mode to a just opened wallet.

promag replied:
But this is the same as just calling `walletView->setPrivacy(gui->isPrivacyModeActivated())`.

hebasto replied:
> But this is the same as just calling `walletView->setPrivacy(gui->isPrivacyModeActivated())`.

Indeed :)
Thank you!

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/16432#issuecomment-633592495).

hebasto commented:
Updated 206d17e4d7131179e7c5a5e74632806859d7d493 -> ad5103667d0dc0a8117a2dd99032ecba2107cdd5 ([pr16432.18](https://github.com/hebasto/bitcoin/commits/pr16432.18) -> [pr16432.19](https://github.com/hebasto/bitcoin/commits/pr16432.19), [diff](https://github.com/hebasto/bitcoin/compare/pr16432.18..pr16432.19)):

- addressed @promag's [comment](https://github.com/bitcoin/bitcoin/pull/16432#discussion_r429690630):
> This condition looks unnecessary/unrelated, could add a comment maybe?

hebasto commented:
Updated ad5103667d0dc0a8117a2dd99032ecba2107cdd5 -> 31be796e44aff80510fb561cc0e8dbd2aa302080 ([pr16432.19](https://github.com/hebasto/bitcoin/commits/pr16432.19) -> [pr16432.20](https://github.com/hebasto/bitcoin/commits/pr16432.20), [diff](https://github.com/hebasto/bitcoin/compare/pr16432.19..pr16432.20)):

- addressed @promag's [comment](https://github.com/bitcoin/bitcoin/pull/16432#discussion_r429805750):
> Why singleShot?
> But this is the same as just calling `walletView->setPrivacy(gui->isPrivacyModeActivated())`.

promag reviewed (commented):
Tested ACK 31be796e44aff80510fb561cc0e8dbd2aa302080.

I still think this could have a different implementation, but this PR is already long. Improvements can be made separately so I'd say this could be merged considering we are in early 0.21.

jonatack commented on src/qt/overviewpage.cpp:
```diff
@@ -152,6 +154,21 @@ void OverviewPage::handleOutOfSyncWarningClicks()
+
+    ui->listTransactions->setVisible(!m_privacy);
+
+    const QString status_tip = m_privacy ? tr("PRIVACY mode is activated. To unmask the values, uncheck Menu->Settings->Mask values.") : "";
```
```suggestion
    const QString status_tip = m_privacy ? tr("Privacy mode is activated for the Overview tab. To unmask the values, uncheck Menu->Settings->Mask values.") : "";
```

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/16432#issuecomment-634041616).

jonatack reviewed (commented):
Tested re-ACK 31be796 with suggestion (here or in a follow-up) to not display "PRIVACY mode" in all caps, and to add "for the Overview tab" to avoid confusion per the principle of least surprise until it is extended to the other tabs as well.

hebasto commented on src/qt/bitcoinunits.cpp:
```diff
@@ -150,6 +153,17 @@ QString BitcoinUnits::formatHtmlWithUnit(int unit, const CAmount& amount, bool p
 
+QString BitcoinUnits::formatWithPrivacy(int unit, const CAmount& amount, SeparatorStyle separators, bool privacy)
+{
+    assert(amount >= 0);
```
> [73d8ef7](https://github.com/bitcoin/bitcoin/commit/73d8ef72742ab9193e9e95158b26176bfaab3f66)
> 
> nit, any reason for this assertion in this function? Looks like it should be in `format()`?

`BitcoinUnits::format()` handles a negative `CAmount` argument correctly. But the code of `BitcoinUnits::formatWithPrivacy()` is not designed to work with the minus sign. This is a deliberate simplification.

hebasto commented:
Updated 31be796e44aff80510fb561cc0e8dbd2aa302080 -> 908b01457bc8454c78fb4d12b7b0d37030f2a6bd ([pr16432.20](https://github.com/hebasto/bitcoin/commits/pr16432.20) -> [pr16432.21](https://github.com/hebasto/bitcoin/commits/pr16432.21), [diff](https://github.com/hebasto/bitcoin/compare/pr16432.20..pr16432.21)):

- addressed @promag's [comment](https://github.com/bitcoin/bitcoin/pull/16432#discussion_r429973527):
> ATM `m_mask_values_action` is always instanced but only added to the GUI when `walletFrame` is set. But I'd make the instancing optional and here I'd `assert(m_mask_values_action)` - it should only be called when wallets are enabled.

- addressed @jonatack's [comment](https://github.com/bitcoin/bitcoin/pull/16432#pullrequestreview-418127354)

> Tested re-ACK [31be796](https://github.com/bitcoin/bitcoin/commit/31be796e44aff80510fb561cc0e8dbd2aa302080) with suggestion (here or in a follow-up) to not display "PRIVACY mode" in all caps, and to add "for the Overview tab" to avoid confusion per the principle of least surprise until it is extended to the other tabs as well.

hebasto commented on src/qt/bitcoingui.cpp:
```diff
@@ -1398,6 +1407,11 @@ void BitcoinGUI::unsubscribeFromCoreSignals()
 
+bool BitcoinGUI::isPrivacyModeActivated() const
+{
+    return m_mask_values_action && m_mask_values_action->isChecked();
```
[Updated](https://github.com/bitcoin/bitcoin/pull/16432#issuecomment-634041616).

jonatack commented:
Tested re-ACK 908b01457bc8454c78fb4d12b7b0d37030f2a6bd only changes are addressing feedback (thanks!) as per `git diff 31be796 908b014`

<details><summary><code>git diff 31be796 908b014</code></summary><p>

```diff
((HEAD detached at origin/pr/16432))$ git diff 31be796 908b014
diff --git a/src/qt/bitcoingui.cpp b/src/qt/bitcoingui.cpp
index a2fd231c05..321812c1f0 100644
--- a/src/qt/bitcoingui.cpp
+++ b/src/qt/bitcoingui.cpp
@@ -1409,7 +1409,8 @@ void BitcoinGUI::unsubscribeFromCoreSignals()
 
 bool BitcoinGUI::isPrivacyModeActivated() const
 {
-    return m_mask_values_action && m_mask_values_action->isChecked();
+    assert(m_mask_values_action);
+    return m_mask_values_action->isChecked();
 }
 
 UnitDisplayStatusBarControl::UnitDisplayStatusBarControl(const PlatformStyle *platformStyle) :
diff --git a/src/qt/overviewpage.cpp b/src/qt/overviewpage.cpp
index cc521c2acb..51682fe2fc 100644
--- a/src/qt/overviewpage.cpp
+++ b/src/qt/overviewpage.cpp
@@ -163,7 +163,7 @@ void OverviewPage::setPrivacy(bool privacy)
 
     ui->listTransactions->setVisible(!m_privacy);
 
-    const QString status_tip = m_privacy ? tr("PRIVACY mode is activated. To unmask the values, uncheck Menu->Settings->Mask values.") : "";
+    const QString status_tip = m_privacy ? tr("Privacy mode is activated for the Overview tab. To unmask the values, uncheck Menu->Settings->Mask values.") : "";
     setStatusTip(status_tip);
     QStatusTipEvent event(status_tip);
     QApplication::sendEvent(this, &event);
```
</p></details>

Apologetic nit: I belatedly noticed that this change makes the message long enough to be potentially cut off at the end (on my display on Debian). Do we need to write `Menu->`? If you retouch, maybe consider something like the following. I tested that it fits in the smallest display size (on my display).
```diff
-    const QString status_tip = m_privacy ? tr("Privacy mode is activated for the Overview tab. To unmask the values, uncheck Menu->Settings->Mask values.") : "";
+    const QString status_tip = m_privacy ? tr("Privacy mode activated for the Overview tab. To unmask the values, uncheck Settings->Mask values.") : "";
```

hebasto commented:
Updated 908b01457bc8454c78fb4d12b7b0d37030f2a6bd -> 8d75115844baefe5cad4d82ec8dce001dd16eb9c ([pr16432.21](https://github.com/hebasto/bitcoin/commits/pr16432.21) -> [pr16432.22](https://github.com/hebasto/bitcoin/commits/pr16432.22), [diff](https://github.com/hebasto/bitcoin/compare/pr16432.21..pr16432.22)):

- addressed @jonatack's [comment](https://github.com/bitcoin/bitcoin/pull/16432#issuecomment-634522397):

> Apologetic nit: I belatedly noticed that this change makes the message long enough to be potentially cut off at the end (on my display on Debian). Do we need to write `Menu->`? If you retouch, maybe consider something like the following. I tested that it fits in the smallest display size (on my display).
> 
> ```diff
> -    const QString status_tip = m_privacy ? tr("Privacy mode is activated for the Overview tab. To unmask the values, uncheck Menu->Settings->Mask values.") : "";
> +    const QString status_tip = m_privacy ? tr("Privacy mode activated for the Overview tab. To unmask the values, uncheck Settings->Mask values.") : "";
> ```

jonatack commented:
Tested ACK 8d75115

Changes since last review:
<details><summary><code>git diff 908b014 8d75115</code></summary><p>

```diff
((HEAD detached at origin/pr/16432))$ git diff 31be796 908b014
$ git diff 908b014 8d75115
diff --git a/src/qt/overviewpage.cpp b/src/qt/overviewpage.cpp
index 51682fe2fc..1ab30fefc7 100644
--- a/src/qt/overviewpage.cpp
+++ b/src/qt/overviewpage.cpp
@@ -163,7 +163,7 @@ void OverviewPage::setPrivacy(bool privacy)
 
     ui->listTransactions->setVisible(!m_privacy);
 
-    const QString status_tip = m_privacy ? tr("Privacy mode is activated for the Overview tab. To unmask the values, uncheck Menu->Settings->Mask values.") : "";
+    const QString status_tip = m_privacy ? tr("Privacy mode activated for the Overview tab. To unmask the values, uncheck Settings->Mask values.") : "";
```
</p></details>

Thanks @hebasto :trophy:

luke-jr commented:
>Apologetic nit: I belatedly noticed that this change makes the message long enough to be potentially cut off at the end (on my display on Debian).

We shouldn't be assuming or breaking with any window or font size... Is this a bug?

jonatack commented:
@luke-jr Good question. My description was imprecise. When the GUI window was at its smallest size, the message wasn't truncated per se but required navigating horizontally using the window scroll bar to see the last word.

jonasschnelli commented:
Tested ACK 8d75115844baefe5cad4d82ec8dce001dd16eb9c

BenWestgate commented:
Current Behavior: "Mask values" mode is off each launch and must be manually re-enabled.
If any wallets load on startup from `settings.json` this leaves user vulnerable to shoulder surfing.

Solution: This mode should persist across restarts of bitcoin-qt by being saved in `Bitcoin-Qt.conf`.
Exactly how `nDisplayUnit=0` will persist the unit to show amounts in.

Expected Behavior: If I shutdown Bitcoin core with it on, I'd expect it to be on when I restart and vice versa.
